atp/client.go

```diff
@@ -328,12 +367,88 @@ func (c *client) sendExecutionResult(runID string, result ExecutionResult) {
 
 func (c *client) sendErrorToAll(err error) {
 	result := NewErrorExecutionResult(err)
-	for runID := range c.runningStepResultChannels {
+	c.mutex.Lock()
+	for runID := range c.runningStepResultEntries {
 		c.sendExecutionResult(runID, result)
 	}
+	c.mutex.Unlock()
+}
+
+func (c *client) handleWorkDoneMessage(runtimeMessage DecodedRuntimeMessage) {
+	var doneMessage WorkDoneMessage
+	var result ExecutionResult
+	if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &doneMessage); err != nil {
+		c.logger.Errorf("Failed to decode work done message (%v) for run ID '%s' ", err, runtimeMessage.RunID)
+		result = NewErrorExecutionResult(fmt.Errorf("failed to decode work done message (%w)", err))
+	} else {
+		result = c.processWorkDone(runtimeMessage.RunID, doneMessage)
+	}
+	c.mutex.Lock()
+	c.sendExecutionResult(runtimeMessage.RunID, result)
+	c.mutex.Unlock()
+}
+
+func (c *client) handleSignalMessage(runtimeMessage DecodedRuntimeMessage) {
+	var signalMessage SignalMessage
+	if err := cbor.Unmarshal(runtimeMessage.RawMessageData, &signalMessage); err != nil {
+		c.logger.Errorf("ATP client for run ID '%s' failed to decode signal message: %v",
+			runtimeMessage.RunID, err)
+		return
+	}
+	c.mutex.Lock()
+	signalChannel, found := c.runningStepEmittedSignalChannels[runtimeMessage.RunID]
+	c.mutex.Unlock()
+	if !found {
+		c.logger.Warningf(
+			"Step with run ID '%s' sent signal '%s'. Ignoring; signal handling is not implemented "+
+				"(emittedSignals is nil).",
+			runtimeMessage.RunID, signalMessage.SignalID)
+		return
+	}
+	c.logger.Debugf("Got signal from step with run ID '%s' with ID '%s'", runtimeMessage.RunID,
+		signalMessage.SignalID)
+	signalChannel <- signalMessage.ToInput(runtimeMessage.RunID)
```

Comment:
Continuing [an earlier conversation](https://github.com/arcalot/arcaflow-plugin-sdk-go/pull/92#discussion_r1661570262), once we unlock the lock at line 400, I think it's possibly for the channel referenced by `signalChannel` to then be closed at line 364, in which case we'll panic at line 410.

Reply:
The simplest way to force ordering in that case is to delay the release of the lock until after we use it. I made that change. I'm not sure if the overhead of an alternative design that keeps track of which channel references aren't done yet would be any less than a single channel send.